src/pdn/src/power_cells.cpp

```diff
@@ -0,0 +1,578 @@
+  }
+}
+
+const std::string GridSwitchedPower::toString(NetworkType type)
+{
+  switch (type) {
+  case STAR:
+    return "STAR";
+  case DAISY:
+    return "DAISY";
+  }
+  return "unknown";
+}
+
+GridSwitchedPower::NetworkType GridSwitchedPower::fromString(const std::string& type, utl::Logger* logger)
+{
+  if (type == "STAR") {
+    return STAR;
+  }
+  if (type == "DAISY") {
+    return DAISY;
+  }
+
+  logger->error(utl::PDN, 197, "Unrecognized network type: {}", type);
+  return STAR;
+}
+
+void GridSwitchedPower::report() const
+{
+  auto* logger = grid_->getLogger();
+  logger->info(utl::PDN, 210, "Switched power cell: {}", cell_->getName());
+  logger->info(utl::PDN, 211, "  Control net: {}", control_->getName());
+  logger->info(utl::PDN, 212, "  Network type: {}", toString(network_));
+}
+
+void GridSwitchedPower::build()
+{
+  if (!insts_.empty()) {
+    // power switches already built and need to be ripped up to try again
+    return;
+  }
+
+  InstTree exisiting_insts;
+  for (auto* inst : grid_->getBlock()->getInsts()) {
+    if (!inst->getPlacementStatus().isFixed()) {
+      continue;
+    }
+
+    odb::Rect bbox;
+    inst->getBBox()->getBox(bbox);
+
+    exisiting_insts.insert({Shape::rectToBox(bbox), inst});
+  }
+
+  auto* target = getLowestStrap();
+  if (target == nullptr) {
+    grid_->getLogger()->error(utl::PDN, 220, "Unable to find a strap to connect power switched to.");
+  }
+
+  odb::dbNet* switched = grid_->getDomain()->getSwitchedPower();
+  odb::dbNet* alwayson = grid_->getDomain()->getAlwaysOnPower();
+  odb::dbNet* ground = grid_->getDomain()->getGround();
+
+  const auto& target_shapes = target->getShapes();
+  ShapeTree targets;
+  for (const auto& [box, shape] : target_shapes.at(target->getLayer())) {
+    if (shape->getNet() != alwayson) {
+      continue;
+    }
+    targets.insert({box, shape});
+  }
+
+  const auto rows = grid_->getDomain()->getRows();
+  bgi::rtree<std::pair<Box, odb::dbRow*>, bgi::quadratic<16>> row_search;
+  for (auto* row : rows) {
+    odb::Rect bbox;
+    row->getBBox(bbox);
+
+    row_search.insert({Shape::rectToBox(bbox), row});
+  }
+  auto get_instance_rows = [&row_search] (odb::dbInst* inst) -> std::set<odb::dbRow*> {
+    std::set<odb::dbRow*> rows;
+
+    odb::Rect box;
+    inst->getBBox()->getBox(box);
+
+    for (auto itr = row_search.qbegin(bgi::intersects(Shape::rectToBox(box)));
+         itr != row_search.qend();
+         itr++) {
+      auto* row = itr->second;
+      odb::Rect row_box;
+      row->getBBox(row_box);
+
+      const auto overlap = row_box.intersect(box);
+      if (overlap.minDXDY() != 0) {
+        rows.insert(row);
+      }
+    }
+
+    return rows;
+  };
+
+  for (auto* row : rows) {
+    const int site_width = row->getSite()->getWidth();
+    cell_->populateAlwaysOnPinPositions(site_width);
+    if (row->getOrient() == odb::dbOrientType::R0) {
+      continue;
+    }
```

Comment:
why?

Reply:
A comment would be helpful

Reply:
added a comment and moved this to the power cell class for clarity